cpp/src/experimental/louvain.cuh

```diff
@@ -33,21 +365,1567 @@ class Louvain {
+
+    std::cout << "rank: " << rank_ << " entering louvain" << std::endl;
+
+    CUDA_TRY(cudaStreamSynchronize(stream_));
+
+    std::cout << "rank: " << rank_ << " computing total_edge_weight" << std::endl;
+    weight_t total_edge_weight;
+    total_edge_weight = experimental::transform_reduce_e(
+      handle_,
+      current_graph_view_,
+      thrust::make_constant_iterator(0),
+      thrust::make_constant_iterator(0),
+      [] __device__(auto, auto, weight_t wt, auto, auto) { return wt; },
+      weight_t{0});
+
+    weight_t best_modularity = weight_t{-1};
+
+    std::cout << "rank: " << rank_ << " total_edge_weight = " << total_edge_weight << std::endl;
+
+    //
+    //  Initialize every cluster to reference each vertex to itself
+    //
+    thrust::sequence(rmm::exec_policy(stream_)->on(stream_),
+                     cluster_v_.begin(),
+                     cluster_v_.end(),
+                     base_vertex_id_);
+    thrust::copy(
+      rmm::exec_policy(stream_)->on(stream_), cluster_v_.begin(), cluster_v_.end(), d_cluster_vec);
+
+    while (num_level < max_level) {
+      compute_vertex_and_cluster_weights();
+
+      weight_t new_Q = update_clustering(total_edge_weight, resolution);
+
+#ifdef DEBUG
+      std::cout << "new_Q = " << new_Q << std::endl;
+#endif
+
+      if (new_Q <= best_modularity) { break; }
+
+      best_modularity = new_Q;
+
+      shrink_graph(d_cluster_vec);
+
+      num_level++;
+    }
+
+    timer_display(std::cout);
+
+#ifdef DEBUG
+    print_v("resulting cluster", d_cluster_vec, number_of_vertices_);
+#endif
+
+    return std::make_pair(num_level, best_modularity);
+  }
+
+ protected:
+  void timer_start(std::string const &region)
+  {
+#ifdef TIMING
+    if (rank_ == 0) hr_timer_.start(region);
+#endif
+  }
+
+  void timer_stop(cudaStream_t stream)
+  {
+#ifdef TIMING
+    if (rank_ == 0) {
+      CUDA_TRY(cudaStreamSynchronize(stream));
+      hr_timer_.stop();
+    }
+#endif
+  }
+
+  void timer_display(std::ostream &os)
+  {
+#ifdef TIMING
+    if (rank_ == 0) hr_timer_.display(os);
+#endif
+  }
+
+ public:
+  // TODO CHECK:  Should be right, as long as
+  //
+  //    src_cluster_cache_v_
+  //    dst_cluster_cache_v_
+  //
+  // are populated properly
+  //
+  weight_t modularity(weight_t total_edge_weight, weight_t resolution)
+  {
+    weight_t Q = -resolution / (total_edge_weight * total_edge_weight) *
+                 experimental::transform_reduce_v(
+                   handle_,
+                   current_graph_view_,
+                   cluster_weights_v_.begin(),
+                   [] __device__(auto p) { return p * p; },
+                   weight_t{0});
+
+#if 1
+    std::cout << "** Q (subtraction) = " << Q << std::endl;
+
+    print_v("weights", vertex_weights_v_);
+    print_v("src_cluster_cache", src_cluster_cache_v_);
+    print_v("dst_cluster_cache", dst_cluster_cache_v_);
+#endif
+
+    rmm::device_vector<weight_t> increase_v(local_num_vertices_, weight_t{0.0});
```

Comment:
Just FYI, here, initialization is unnecessary, and in such case,
```
rmm::device_uvector<weight_t> increase_v(local_num_vertices_, handle_.get_stream());
```
can save initialization time.

You can exploit this later if you want to do some extensive performance tuning.